PolyEngine/Core/Src/pe/core/storage/Queue.hpp

```diff
@@ -78,198 +77,337 @@ namespace pe::core::storage
+		private:
+			ReverseIterator(Queue<T>* queue, size_t index) : q(queue), idx(index) {};
+			Queue<T>* q;
+			size_t idx;
+		};
+
+		class ConstReverseIterator final : public BaseObjectLiteralType<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ConstReverseIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ConstReverseIterator& rhs) const { return idx != rhs.idx; }
+			
+			const T& operator*() const { return q->m_data[idx]; }
+			const T* operator->() const { return q->m_data[idx]; }
+
+			ConstReverseIterator& operator++() { idx = q->getPrevIdx(idx); return *this; }
+			ConstReverseIterator operator++(int) { ConstReverseIterator ret(q, idx); idx = q->getPrevIdx(idx); return ret; }
+			ConstReverseIterator& operator--() { idx = q->getNextIdx(idx); return *this; }
+			ConstReverseIterator operator--(int) { ConstReverseIterator ret(q, idx); idx = q->getNextIdx(idx); return ret; }
+		private:
+			ConstReverseIterator(const Queue<T>* queue, size_t index) : q(queue), idx(index) {};
+			const Queue<T>* q;
+			size_t idx;
+		};
+
+		Iterator begin() { return Iterator(this, m_head); }
+		Iterator end() { return Iterator(this, m_tail); }
+		ConstIterator cbegin() const { return ConstIterator(this, m_head); }
+		ConstIterator cend() const { return ConstIterator(this, m_tail); }
+		ReverseIterator rbegin() { return ReverseIterator(this, m_tail ? m_tail - 1 : m_tail); }
+		ReverseIterator rend() { return ReverseIterator(this, m_tail); }
+		ConstReverseIterator crbegin() const { return ConstReverseIterator(this, m_tail ? m_tail - 1 : m_tail); }
+		ConstReverseIterator crend() const { return ConstReverseIterator(this, m_tail); }
 
 	private:
 		//------------------------------------------------------------------------------
-		size_t GetPrevIdx(size_t idx) const { return (idx == 0) ? Capacity - 1 : idx - 1; }
-		size_t GetNextIdx(size_t idx) const { return (idx + 1) % Capacity; }
-		void DecreaseIdx(size_t& idx) const { idx = GetPrevIdx(idx); }
-		void AdvanceIdx(size_t& idx) const { idx = GetNextIdx(idx); }
-		size_t GetNthIdx(size_t& n)	const { return (Head + n) % Capacity; }
-
-		//------------------------------------------------------------------------------
-		size_t FindIdx(const T& rhs) const
+		size_t getPrevIdx(size_t idx) const
 		{
-			for (size_t i = 0; i < Size; ++i)
-				if (Data[GetNthIdx(i)] == rhs)
-					return GetNthIdx(i);
-
-			return GetSize();
+			// PhysicalSize takes guard cell into account
+			const size_t physicalSize = m_size + 1;
+			idx = (idx + physicalSize - 1 - m_head) % physicalSize;
+			return idx + m_head;
+		}
+		
+		size_t getNextIdx(size_t idx) const
+		{
+			// PhysicalSize takes guard cell into account
+			const size_t physicalSize = m_size + 1;
+			idx = (idx + 1 - m_head) % physicalSize;
```

Comment:
we can still get an overflow here if the head is bigger than idx + 1. Add physicalSize to idx as is done in getPrevIdx.